README.md
koa-github [![Build Status](https://secure.travis-ci.org/dead-horse/koa-github.png)](http://travis-ci.org/dead-horse/koa-github)
==========

simple github auth middleware for koa

[![NPM](https://nodei.co/npm/koa-github.png?downloads=true)](https://nodei.co/npm/koa-github/)

## Example

```
//use http://localhost:7001 to test

var koa = require('koa');
var http = require('http');
var session = require('koa-sess');
var githubAuth = require('koa-github');

var app = koa();

app.name = 'nae-web';
app.keys = ['key1', 'key2'];

app.use(session());
app.use(githubAuth({
  clientID: '5ec1d25d2a3baf99a03c',
  clientSecret: '513607494a244e2759738cae3d50a89494c1e7f0',
  callbackURL: 'http://localhost:7001',
  userKey: 'user',
  timeout: 10000
}));

app.use(function *handler() {
  if (!this.session.githubToken) {
    this.body = '<a href="/github/auth">login with github</a>';
  } else {
    this.body = this.session.user;
  }
});

app.on('error', function (err) {
  if (!err.status || err.status >= 500) {
    logger.error(err);
  }
});

http.createServer(app.callback()).listen(7001);
```

## Options

```
  @param {Object} options
    - [String] clientID      github client ID     // regist in https://github.com/settings/applications
    - [String] clientSecret  github client secret
    - [String] callbackURL   github redirect url
    - [String] signinPath    sign in with github's triggle path, default is `/github/auth`
    - [String] tokenKey      session key, default is githubToken
    - [String] userKey       user key, if set user key, will request github once to get the user info
    - [Array]  scope         A comma separated list of scopes
    - [Number] timeout       request github api timeout
    - [String] redirect      redirect key when call signinPath, so we can redirect after auth, default is `redirect_uri`
```

* clientID, clentSecret and callbackURL are regist in https://github.com/settings/applications.
* if you set userKey field, `koa-github` will request to get the user info and set this object to `this.session[options.userKey]`, otherwise `koa-github` will not do this request.
* if you triggle by `/github/auth?redirect_uri=/callback`, `koa-github` will redirect to `/callback` after auth, and the redirect_uri only accept start with `/`.

## Licences
(The MIT License)

Copyright (c) 2013 dead-horse and other contributors

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

0.0.5 / 2014-03-12
==================

  * use istanbul, remove unused dependencies
  * Merge pull request #1 from zdalniacy/master
  * Merge pull request #1 from zdalniacy/patch-1
  * Fixing coveralls for koa-github

0.0.4 / 2014-01-08
==================

  * use utility
  * add travis.yml
  * add bundge

0.0.3 / 2014-01-02
==================

  * support redirect
  * remove redirectPath
  * update denpendencies
  * remove devDependencies co
  * add test
  * use assert
  * add authors

0.0.2 / 2014-01-01
==================

  * add readme

0.0.1 / 2014-01-01
==================

  * add example
  * add github auth koa middleware
  * Initial commit


test/support/server.js
//use http://localhost:7001 to test

var koa = require('koa');
var http = require('http');
var session = require('koa-sess');
var githubAuth = require('../../');

var app = koa();

app.name = 'nae-web';
app.keys = ['key1', 'key2'];

app.use(session());
app.use(githubAuth({
  clientID: '5ec1d25d2a3baf99a03c',
  clientSecret: '513607494a244e2759738cae3d50a89494c1e7f0',
  callbackURL: 'http://localhost:7001/github/auth/callback',
  userKey: 'user',
  timeout: 10000
}));

app.use(function *handler() {
  if (!this.session.githubToken) {
    this.body = '<a href="/github/auth">login with github</a>';
  } else {
    this.body = this.session.user;
  }
});

app.on('error', function (err) {
  if (!err.status || err.status >= 500) {
    logger.error(err);
  }
});

module.exports = http.createServer(app.callback());


test/koa-github.test.js
/**!
 * koa-github - test/koa-github.test.js
 *
 * MIT Licensed
 *
 * Authors:
 *   dead_horse <dead_horse@qq.com> (http://deadhorse.me)
 */

'use strict';

/**
 * Module dependencies.
 */

var app = require('./support/server');
var should = require('should');
var request = require('supertest');

describe('koa-github', function () {
  describe('GET /github/auth', function () {
    it('should redirect ok', function (done) {
      request(app)
      .get('/github/auth')
      .expect(302, done);
    });
  });

  describe('GET /github/auth/callback', function () {
    it('should 400', function (done) {
      request(app)
      .get('/github/auth/callback?code=123')
      .expect(400, done);
    });

    it('should 403', function (done) {
      request(app)
      .get('/github/auth/callback?code=123&state=123')
      .expect(403, done);
    });
  });
});

index.js
/**!
 * koa-github - index.js
 *
 * MIT Licensed
 *
 * Authors:
 *   dead_horse <dead_horse@qq.com> (http://deadhorse.me)
 */

'use strict';

/**
 * Module dependencies.
 */

var urllib = require('co-urllib');
var debug = require('debug')('koa-github');
var utility = require('utility');
var util = require('util');
var qsParse = require('querystring').parse;
var urlParse = require('url').parse;
var assert = require('assert');

var defaultOptions = {
  tokenKey: 'githubToken',
  signinPath: '/github/auth',
  timeout: 5000,
  scope: ['user'],
  redirect: 'redirect_uri'
};

/**
 * auth with github
 * need use session middleware before
 * see http://developer.github.com/v3/oauth/#web-application-flow
 *
 * @param {Object} options
 *   - [String] clientID      github client ID
 *   - [String] clientSecret  github client secret
 *   - [String] callbackURL   github callback url
 *   - [String] signinPath    sign in with github's triggle path, default is /github/auth
 *   - [String] tokenKey      session key, default is githubToken
 *   - [String] userKey       user key, if set user key, will request github once to get the user info
 *   - [Array]  scope         A comma separated list of scopes
 *   - [Number] timeout       request github api timeout
 *   - [String] redirect      redirect key when call signinPath, so we can redirect after auth, default is redirect_uri
 *
 */
module.exports = function (options) {
  options = options || {};
  if (!options.clientID || !options.clientSecret || !options.callbackURL) {
    throw new Error('github auth need clientID, clientSecret and callbackURL');
  }
  for (var key in defaultOptions) {
    if (!utility.has(options, key)) {
      options[key] = defaultOptions[key];
    }
  }
  options.callbackURL = options.callbackURL;
  options.callbackPath = urlParse(options.callbackURL).path;

  urllib.TIMEOUT = options.timeout;
  debug('init github auth middleware with options %j', options);

  return function *githubAuth(next) {
    if (!this.session) {
      return this.throw('github auth need session', 500);
    }

    // first step: redirect to github
    if (this.path === options.signinPath) {
      var state = utility.randomString();
      var redirectUrl = 'https://github.com/login/oauth/authorize?';
      redirectUrl = util.format('%sclient_id=%s&redirect_uri=%s&scope=%s&state=%s',
        redirectUrl, options.clientID, options.callbackURL, options.scope, state);

      this.session._githubstate = state;

      //try to get the redirect url and set it to session
      try {
        var redirect = decodeURIComponent(urlParse(this.url, true).query[options.redirect] || '');
        if (redirect[0] === '/') {
          this.session._githubredirect = redirect;
          debug('get github callback redirect uri: %s', redirect);
        }
      } catch (err) {
        debug('decode redirect uri error');
      }
      debug('request github auth, redirect to %s', redirectUrl);
      //if already signin
      if (this.session[options.tokenKey]) {
        debug('already has github token');
        redirectUrl = this.session._githubredirect || '/';
        delete this.session._githubredirect;
      }
      return this.redirect(redirectUrl);
    }

    // secound step: github callback
    if (this.path === options.callbackPath) {
      //if already signin
      if (this.session[options.tokenKey]) {
        debug('already has github token');
        return this.redirect('/');
      }

      debug('after auth, jump from github.');
      var url = urlParse(this.request.url, true);

      // must have code
      if (!url.query.code || !url.query.state) {
        debug('request url need `code` and `state`');
        return this.throw(400);
      }

      // check the state, protect against cross-site request forgery attacks
      if (url.query.state !== this.session._githubstate) {
        debug('request state is %s, but the state in session is %s',
          url.query.state, this.session._githubstate);
        delete this.session._githubstate;
        return this.throw(403);
      }

      //step three: request to get the access token
      var tokenUrl = 'https://github.com/login/oauth/access_token';
      var requsetOptions = {
        data: {
          client_id: options.clientID,
          client_secret: options.clientSecret,
          code: url.query.code
        }
      };
      debug('request the access token with data: %j', requsetOptions.data);
      var token;
      try {
        var result = yield urllib.request(tokenUrl, requsetOptions);
        assert.equal(result[1].statusCode, 200,
          'response status ' + result[1].statusCode + ' not match 200');

        token = qsParse(result[0].toString()).access_token;
        assert(token, 'response without access_token');
      } catch (err) {
        return this.throw('request github token error: ' + err.message, 500);
      }

      this.session[options.tokenKey] = token;
      debug('get access_token %s and store in session.%s', token, options.tokenKey);
      delete this.session._githubstate;

      //step four: if set userKey, get user
      if (options.userKey) {
        var result;
        try {
          var userUrl = 'https://api.github.com/user';
          var authOptions = {
            headers: {
              Authorization: 'token ' + token,
              'user-agent': 'koa-github'
            },
            dataType: 'json'
          };
          result = yield urllib.request(userUrl, authOptions);
          assert.equal(result[1].statusCode, 200,
            'response status ' + result[1].statusCode + ' not match 200');
          assert(result[0], 'response without user info');
        } catch (err) {
          return this.throw('request github user info error: ' + err.message, 500);
        }
        debug('get user info %j and store in session.%s', result[0], options.userKey);
        this.session[options.userKey] = result[0];
      }
      var githubredirect = this.session._githubredirect || '/';
      delete this.session._githubredirect;
      return this.redirect(githubredirect);
    }

    yield next;
  };
};


example.js
//use http://localhost:7001 to test

var koa = require('koa');
var http = require('http');
var session = require('koa-sess');
var githubAuth = require('./');

var app = koa();

app.name = 'nae-web';
app.keys = ['key1', 'key2'];

app.use(session());
app.use(githubAuth({
  clientID: '5ec1d25d2a3baf99a03c',
  clientSecret: '513607494a244e2759738cae3d50a89494c1e7f0',
  callbackURL: 'http://localhost:7001/github/auth/callback',
  userKey: 'user',
  timeout: 10000
}));

app.use(function *handler() {
  if (!this.session.githubToken) {
    this.body = '<a href="/github/auth?redirect_uri=/callback">login with github</a>';
  } else {
    this.body = this.session.user;
  }
});

app.on('error', function (err) {
  if (!err.status || err.status >= 500) {
    logger.error(err);
  }
});

http.createServer(app.callback()).listen(7001);
console.log('open http://localhost:7001/ in your browser!');
